proxy/proxy.go

```diff
@@ -960,42 +999,40 @@ func (p *Proxy) makeBackendRequest(ctx *context, requestContext stdlibcontext.Co
 		setTag(ctx.proxySpan, HTTPUrlTag, u.String())
 	p.setCommonSpanInfo(u, req, ctx.proxySpan)
 
-	carrier := ot.HTTPHeadersCarrier(req.Header)
-	_ = p.tracing.tracer.Inject(ctx.proxySpan.Context(), ot.HTTPHeaders, carrier)
-
-	req = req.WithContext(ot.ContextWithSpan(req.Context(), ctx.proxySpan))
+	req = tracing.Inject(reqContext, req, ctx.proxySpan, p.tracing.Tracer)
 
 	p.metrics.IncCounter("outgoing." + req.Proto)
-	ctx.proxySpan.LogKV("http_roundtrip", StartEvent)
-	req = injectClientTrace(req, ctx.proxySpan)
+	p.tracing.logEvent(ctx.proxySpan, "http_roundtrip", StartEvent)
+	req = p.injectClientTrace(req, ctx.proxySpan)
 
 	response, err := roundTripper.RoundTrip(req)
 	if endpointMetrics != nil {
 		endpointMetrics.IncRequests(routing.IncRequestsOptions{FailedRoundTrip: err != nil})
 	}
-	ctx.proxySpan.LogKV("http_roundtrip", EndEvent)
+
+	p.tracing.logEvent(ctx.proxySpan, "http_roundtrip", EndEvent)
 	if err != nil {
 		if errors.Is(err, skpio.ErrBlocked) {
-			p.tracing.setTag(ctx.proxySpan, BlockTag, true)
-			p.tracing.setTag(ctx.proxySpan, HTTPStatusCodeTag, uint16(http.StatusBadRequest))
+			p.tracing.setTag(ctx.proxySpan, BlockTag, "true")
+			p.tracing.setTag(ctx.proxySpan, HTTPStatusCodeTag, fmt.Sprint(uint16(http.StatusBadRequest)))
 			return nil, &proxyError{err: err, code: http.StatusBadRequest}
 		}
-		p.tracing.setTag(ctx.proxySpan, ErrorTag, true)
+		p.tracing.setTag(ctx.proxySpan, ErrorTag, "true")
 
 		// Check if the request has been cancelled or timed out
 		// The roundtrip error `err` may be different:
 		// - for `Canceled` it could be either the same `context canceled` or `unexpected EOF` (net.OpError)
 		// - for `DeadlineExceeded` it is net.Error(timeout=true, temporary=true) wrapping this `context deadline exceeded`
-		if cerr := req.Context().Err(); cerr != nil {
-			ctx.proxySpan.LogKV("event", "error", "message", ensureUTF8(cerr.Error()))
+		if cerr := reqContext.Err(); cerr != nil {
+			p.tracing.logErrorEvent(ctx.proxySpan, cerr.Error())
 			if cerr == stdlibcontext.Canceled {
 				return nil, &proxyError{err: cerr, code: 499}
 			} else if cerr == stdlibcontext.DeadlineExceeded {
 				return nil, &proxyError{err: cerr, code: http.StatusGatewayTimeout}
 			}
 		}
 
-		ctx.proxySpan.LogKV("event", "error", "message", ensureUTF8(err.Error()))
+		p.tracing.logErrorEvent(ctx.proxySpan, err.Error())
```

Comment:
are you sure you don't need `ensureUTF8`?
I think we should wrap `cerr.Error()` here